OEL-1227: Move content banner into a block.

donquixote opened the pull request:
[OEL-1227](https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-1227)

donquixote committed:
OEL-1227: Create and place the ContentBannerBlock.

donquixote committed:
OEL-1227: Create 'oe_w_content_banner' view mode.

donquixote committed:
OEL-1227: Configure and theme view modes for 'oe_sc_event', moving fields into content banner.

donquixote committed:
OEL-1227: Noise in teaser view mode for 'oe_sc_event'.

donquixote committed:
OEL-1227: Add the 'banner' region.

donquixote committed:
OEL-1227: Configure and theme view modes for 'oe_sc_news', moving fields into content banner.

donquixote committed:
OEL-1227: Test full pages.

donquixote committed:
OEL-1227: Make 'oe_whitelabel' the default theme in test.

donquixote committed:
OEL-1227: Set up sparql in functional tests, to prevent trouble with blocks.

Why?
We are going to enable 'block' module as a dependency of oe_whitelabel_helper.
As a consequence, all block placements from config/optional will be installed in the test.
At least one of them requires sparql.

Changes:
- Set up sparql in base test class.
- Let CorporateHeaderLogosTest also inherit from the base class.

donquixote committed:
OEL-1227: Also assert event body text.

donquixote committed:
OEL-1227: Use https links in test, to pacify inspection tools.

donquixote committed:
OEL-1227: Code style, comments and minor refactor in tests.

Changes:
- Reorder imports.
- Change line breaks and blank lines.
- Change doc comments on test methods.
- Change inline comments.
- Change order of assertions a little.

Rationale for inline comments and line breaks:
- When rendering the full page, inline comments are somewhat helpful, but not too many.
- Blank lines are nice for structuring content.
- The teaser view mode is simple enough that we don't need inline comments.
- The existing comments in teaser view mode were wrong anyway, because they mentioned "content banner".

donquixote committed:
OEL-1227: Example node as return value instead of object property.

donquixote committed:
OEL-1227: Enable block module in tests.

donquixote committed:
OEL-1227: Assert that content banner is found in page, for better reporting on failure.

donquixote committed:
OEL-1227: Set provider = 'node' in block definition, to force dependency.

brummbar commented on modules/oe_whitelabel_helper/src/Plugin/Block/ContentBannerBlock.php:
```diff
@@ -0,0 +1,50 @@
+    $node = $this->getContextValue('node');
+
+    /** @var \Drupal\Core\Entity\EntityViewModeInterface|null $view_mode_object */
+    $view_mode_object = \Drupal::entityTypeManager()
```
Y U NO ContainerFactoryPluginInterface?

donquixote replied:
Strange, I probably wanted to do this and then I didn't..

brummbar commented on modules/oe_whitelabel_helper/src/Plugin/Block/ContentBannerBlock.php:
```diff
@@ -0,0 +1,50 @@
+      ->getStorage('entity_view_display')
+      ->load('node.' . $node->bundle() . '.oe_w_content_banner');
+
+    return AccessResult::allowedIf($view_mode_object !== NULL);
```
Just to be extra careful, let's add a check if the current user can view the node. Most of the time this will be rendered in the node view page, so the page access altogether will be denied, but it's safer to add the check.

donquixote replied:
I added the access check.

I was wondering if we need to add the node as cacheable dependency. But it seems we don't, because the node is coming from a context, and therefore the caching is already sensitive to that.

I also wondered if we should add a test for the cache tags context. But as there are no cache tags, it would be quite pointless. Also, I did not find any such tests in oe_theme.

brummbar commented on modules/oe_whitelabel_helper/src/Plugin/Block/ContentBannerBlock.php:
```diff
@@ -0,0 +1,50 @@
+  public function build(): array {
+    /** @var \Drupal\Core\Entity\EntityInterface $entity */
+    $entity = $this->getContextValue('node');
+    $view_builder = \Drupal::entityTypeManager()->getViewBuilder($entity->getEntityTypeId());
```
Same as above, dependency injection.

donquixote replied:
done

brummbar commented on oe_whitelabel.info.yml:
```diff
@@ -23,6 +23,7 @@ regions:
   header_top: Header top
   header_left: Header left
   header_right: Header right
+  banner: Banner
```
Maybe we could call this in a more generic way, like "Page header" or "Content top" or something similar.

donquixote replied:
You mean use one of the existing regions?
My goal was to put this outside of `class="container"`  elements, to have natural full width.
It is not fully needed, because the element itself has `100vw`, but if we already move it to a separate region, we can go all the way.

In fact the `width: 100vw` is a bit clumsy, because it does not subtract the width of the vertical scrollbar. So it means we need an `overflow-x: hidden` somewhere.
But as long as the 100vw is in bcl, it does not really matter which region we use..

What's worse, all the regions following this one are within `class="row"`, which messes with the container padding.
https://citnet.tech.ec.europa.eu/CITnet/jira/browse/OEL-1226?jql=project%20%3D%20oel%20AND%20text%20~%20%22container%20padding%20row%22
In our current solution, this required us to add classes from outside to mitigate the problem.

donquixote replied:
One question though is how this should be placed in relation to breadcrumb and tabs.

donquixote replied:
Fyi, the first comment in OEL-1226 has a screenshot that shows a symptom of the 100vw problem. Part of the text for "full-width section for bcl-content-banner" is cut off on the left.

donquixote replied:
A good experiment is to put a block into the sidebar region.
We don't want the banner to show up next to the sidebar.

brummbar replied:
The place where you put the new region is perfect, I didn't like the name much. We had a chat with @drishu and we will follow the naming used in Olivero core theme: `hero: 'Hero (full width)'`

brummbar replied:
The place where you put the new region is perfect, I didn't like the name much. We had a chat with @drishu and we will follow the naming used in Olivero core theme: `hero: 'Hero (full width)'`

donquixote committed:
OEL-1227: Use dependency injection in ContentBannerBlock.

donquixote committed:
OEL-1227: Check node view access in content banner block.

donquixote committed:
OEL-1227: Hard-code entity type = 'node' in content banner block.

donquixote committed:
OEL-1227: Drop local variable, add line break.

donquixote committed:
OEL-1227: Rename theme region 'banner' -> 'hero'.

drishu commented on tests/src/Functional/ContentEventRenderTest.php:
```diff
@@ -91,25 +87,30 @@ protected function setUp(): void {
+
+    // Select the content banner element.
     $content_banner = $crawler->filter('.bcl-content-banner');
+    $this->assertSame(1, $content_banner->count(), 'Content banner found.');
```
```suggestion
    $this->assertSame(1, $content_banner->count(), 'Content banner not found.');
```
the message displays when this fails, same in news pls

donquixote replied:
I found this in Drupal core a lot. But I now learn this is a relic from the past, and should go:
https://www.drupal.org/project/coding_standards/issues/2057905#comment-9470963
So I will change it.

donquixote replied:
After talk with Francesco I am dropping the comment and use assertCount().

drishu commented on modules/oe_whitelabel_helper/oe_whitelabel_helper.info.yml:
```diff
@@ -5,7 +5,13 @@ package: OpenEuropa Whitelabel Theme
+
+config_devel:
+  optional:
+    # A view mode should not depend on a theme, so it is exported in a module.
```
needles to say, as in not needed, lets add comments only where clarifications are needed

donquixote replied:
I have a habit to add a comment whenever I was not sure about something myself and had to look it up or learn it.

donquixote replied:
But yeah let's drop, otherwise we will have comments like this all over the place.

drishu reviewed (changes requested):
Other than these two small points very nice changes.

donquixote committed:
OEL-1227: Drop assertion message, and use assertCount().

Firstly, the assertion message should have been inverted:
See https://www.drupal.org/project/coding_standards/issues/2057905#comment-9470963

Secondly, we avoid assertion messages when possible. People can look up the line number on fail.

donquixote committed:
OEL-1227: Drop comment in info file.

donquixote committed:
OEL-1227: Move view mode config from config/optional to config/install, because other configs depend on it.
